### ISSUE TITLE
Invalid JS: `[a?.b…]` compiles to invalid JavaScript

### ISSUE THREAD

**Issue by maxnordlund** [jashkenas/coffeescript#4260]
I found this today, below is the reduced code to reproduce this error on [coffeescript.org](http://coffeescript.org/#try:handler%20%3D%20%28error%29%20%3D%3E%0A%20%20if%20error%20instanceof%20Error%0A%20%20%20%20arr.push%28error%29%0A%20%20else%0A%20%20%20%20arr.push%28error.data%3F.errors...%29%0A%20%20%40showErrors%20%3D%20!!foo.length%0A).

``` coffeescript
handler = (error) =>
  if error instanceof Error
    arr.push(error)
  else
    arr.push(error.data?.errors...)
  @showErrors = !!foo.length
```

``` javascript
var handler;

handler = (function(_this) {
  return function(error) {
    var ref;
    if (error instanceof Error) {
      arr.push(error);
    } else {
      // The loose if-statement is the error
      arr.push.apply(_this,       if ((ref = error.data) != null) {
        ref.errors;
      });
    }
    return _this.showErrors = !!foo.length;
  };
})(this);
```

**Comment by vendethiel** [comment]
Whoops, good catch...

**Comment by lydell** [comment]
Minimal case:

```
coffee -bpe '[a?.b...]'
var slice = [].slice;

slice.call(  if (typeof a !== "undefined" && a !== null) {
    a.b;
  });
```

**Comment by GeoffreyBooth** [comment]
Either a duplicate or else very closely related to #1349.

**Comment by GeoffreyBooth** [comment]
Fixed by #4644.
